pkg/credentialprovider/plugin/plugin.go

```diff
@@ -176,13 +181,11 @@ func (c *cacheExpirationPolicy) IsExpired(entry *cache.TimestampedEntry) bool {
 // Provide returns a credentialprovider.DockerConfig based on the credentials returned
 // from cache or the exec plugin.
 func (p *pluginProvider) Provide(image string) credentialprovider.DockerConfig {
+
 	if !p.isImageAllowed(image) {
 		return credentialprovider.DockerConfig{}
 	}
 
-	p.Lock()
-	defer p.Unlock()
-
 	cachedConfig, found, err := p.getCachedCredentials(image)
```

Comment:
getCachedCredentials does not look thread-safe (has unprotected reads and writes of `cache` and `lastCachePurge`) ... is there a test that calls `Provide` many times in parallel to exercise race-safety here?

Reply:
also, this function has later access to `p.cache` which is now unprotected... is `p.cache` threadsafe?

Reply:
cache looks thread safe https://github.com/kubernetes/client-go/blob/master/tools/cache/expiration_cache.go#L38
is it fine if i add the `purge` call and  `cache add` call under lock? Making write thread safe.

There was a previous concern having whole provide function under lock could slow it down for parallel image pulls.

There is no test case, I have verified locally. I will add a unit test for race saftey.

Reply:
@liggitt Added test case for parallel call to provide and wrapped `lastCachePurge` in lock, rest cache rw operations are safe, have tested with `go test -race`